src/gui/processing/qgsprocessingalgorithmdialogbase.cpp

```diff
@@ -592,6 +592,15 @@ void QgsProcessingAlgorithmDialogBase::setInfo( const QString &message, bool isE
   processEvents();
 }
 
+void QgsProcessingAlgorithmDialogBase::reject()
+{
+  if ( mMainWidget )
```

Comment:
I'm not sure about this fix -- mMainWidget should already be getting deleted along with the dialog, because it's parented to the dialog in setMainWidget. Possibly what's actually happening here is that the whole dialog is being leaked on cancel, not just the widget.

What happens if you put a

    setAttribute( Qt::WA_DeleteOnClose );

in reject instead?

Reply:
It works smoothly! 
Shall I replace the previous commit by a new one, perhaps calling `setAttribute( Qt::WA_DeleteOnClose );` in the constructor?

Reply:
Not in the constructor. It gets complex, but if you close the dialog WHILE an algorithm is running in the background then we don't want to delete the dialog. Rather we hide it away until the algorithm execution is finished, just in case the user clicks the running task and wants to see the log and detailed progress reports.

So we should only delete the dialog on close while its not running (hence putting this in the reject override)

Reply:
Although -- you better double check the reject **isn't** called while the algorithm is executing...!

Reply:
Alright, then https://github.com/qgis/QGIS/pull/34796/commits/52bf7a07b36570e88d124316ad66d9ff7aef015c sets the `Qt::WA_DeleteOnClose` in reject(), only if there is no task running.

I've tested that this successfully solves the initial issue removing unexpected refactor fields confirmation dialogs.